server/src/com/sixsq/slipstream/ssclj/resources/spec/identification.cljc

```diff
@@ -0,0 +1,29 @@
+(ns com.sixsq.slipstream.ssclj.resources.spec.identification
+  (:require
+    [clojure.spec.alpha :as s]
+    [com.sixsq.slipstream.ssclj.resources.spec.common :as cimi-common]
+    [com.sixsq.slipstream.ssclj.resources.spec.core :as cimi-core]
+    [com.sixsq.slipstream.ssclj.util.spec :as su]))
+
+; {
+;   "user": string,
+; 	"idkey": string,
+; 	"deviceid": string
+; }
+
+
+(s/def :cimi.identification/user ::cimi-core/nonblank-string)
+(s/def :cimi.identification/idkey string?)
+(s/def :cimi.identification/deviceid string?)
```

Comment:
I have followed the other existing resources as an example as best as possible.
In this case, I dont understand your comment, how do I know what? All the files in the folder have this kind of "variable" or "field" definition... for example, this is from the agent resource:

`(s/def :cimi.agent/device_id ::cimi-core/nonblank-string) ; device_id
(s/def :cimi.agent/device_ip string?)
(s/def :cimi.agent/leader_id string?)
(s/def :cimi.agent/leader_ip string?)
(s/def :cimi.agent/authenticated boolean?)
(s/def :cimi.agent/connected boolean?)
(s/def :cimi.agent/isLeader boolean?)
(s/def :cimi.agent/backup_ip string?)
(s/def :cimi.agent/childrenIPs (s/coll-of string?))`

so, I dont understand where is the error here...

Reply:
I mean the `deviceid`, how do you know its value when the user is registering

Reply:
If I am not wrong, according with my definition, that field can be empty but not the idkey. Of course,  when a user is created I wont store anything in the deviceid but only when a new device is registered.

Reply:
that won't work:
 1. Data replication only happens from child to leader and not vice versa, which means the agents will not have access to the cloud agent, where that resource has been stored...unless some other machinery (like the CAU) does the work for the agents
 2. one user can have multiple devices, thus multiple device IDs
 3. that's not a good practice, leaving an optional field just because you don't have it at the time of creation. Better would be to have a different resource. Which we already have: `device`.

Reply:
1. The identification module works directly with the cloud agent to register devices, so, from UPC we expect that at least during the device registration, it has access to the Internet and thus, access to the cloud agent. 
2. I dont see the problem with this point. 
3. Almost all the resources (if not all of them) have optional fields for this or any other reason.

Reply:
1. so what happens during this interaction? where are these agents registered? and moreover, how is the agent authenticating with the cloud agent??... because no credentials are passed to the compose file nor any of the components running at the edge
 2. you've set it as a string, while at least it would need to be a list
 3. an optional field means something that it is not required in order to define a specific resource. Let's say my `physical description` is a CIMI resource...fields like height, weight, hair-color, eye-color, skin-color, etc... would be **required**. But other fields like blood-type, age, has-braces, etc. would be **optional**, because they only provide extra information that helps classifying my physical description. A real example in mf2c is that all resources have `name` and `description` as optional, why? because these fields are purely metadata, and are just there to help make sense of that resource. Another real example is the resource `agreement`, where the field `assessment` is optional. This makes sense. Assessment is related to an agreement, and an agreement might be or not assessed. **Now back to your case**, 1) you know beforehand that deviceID will be critical in the workflow, however you're marking it as optional; 2) deviceID is nowhere representative of the resource `identification` we are handling here...it is not related with identification at all so it should not be a part of this resource; 3) even if there were others doing things for the wrong reasons, that is not an excuse for us to do the same 😛 . So my point here **is**: there is already a `device` resource, created by your colleague Souvik. It already contains a field for `deviceID`. So ideally, you could extend his schema, to also include a field for `idkey` and `user`. This way, because these agent's resource are going to be replicated to the leader, this filter `GET /api/device?$filter=user=user/alex or idkey=<idkey>` will always give you all the devices and respective IDKey that belong to a user